tests/Makefile/MakefileTest.php

```diff
@@ -0,0 +1,303 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Makefile;
+
+use function array_filter;
+use function array_shift;
+use function chdir;
+use function current;
+use function file_get_contents;
+use function getcwd;
+use function implode;
+use PHPUnit\Framework\TestCase;
+use function shell_exec;
+use function sprintf;
+use function strpos;
+use function substr_count;
+
+/**
+ * @coversNothing
+ *
+ * @group infra
+ */
+final class MakefileTest extends TestCase
+{
+    private const MAKEFILE_PATH = __DIR__ . '/../../Makefile';
+
+    /**
+     * @var string
+     */
+    private static $originalCwd;
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function setUpBeforeClass(): void
+    {
+        self::$originalCwd = getcwd();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function tearDown(): void
+    {
+        chdir(self::$originalCwd);
+    }
+
+    public function test_the_default_goal_is_the_help_command(): void
+    {
+        $output = shell_exec(sprintf('make -f %s', self::MAKEFILE_PATH));
+
+        $expectedOutput = <<<'EOF'
+[33mUsage:[0m
+  make TARGET
+
+[32m#
+# Commands
+#---------------------------------------------------------------------------[0m
+
+[33mcompile:[0m	 Bundles Infection into a PHAR
+[33mcs:[0m	  	 Runs PHP-CS-Fixer
+[33mphpstan:[0m  	 Runs PHPStan
+[33manalyze:[0m	 Runs Static analyzers and various other checks
+[33mvalidate:[0m	 Checks that the composer.json file is valid
+[33mtest:[0m		 Runs all the tests
+[33mtest-unit:[0m	 Runs the unit tests
+[33mtest-e2e:[0m 	 Runs the end-to-end tests
+[33mtest-infection:[0m  Runs Infection against itself
+
+EOF;
+
+        $this->assertSame($expectedOutput, $output);
+    }
+
+    public function test_the_makefile_can_be_parsed(): void
+    {
+        (new Parser())->parse(file_get_contents(self::MAKEFILE_PATH));
```

Comment:
Since the parser is stateless unlike what I originally thought it was gonna be, I think `Parser::parse()` can be made static